File: app/src/main/java/me/yxp/qfun/hook/api/OnAIOViewUpdate.java
package me.yxp.qfun.hook.api;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.LinearLayout;

import java.lang.reflect.Method;

import me.yxp.qfun.hook.base.ApiHookItem;
import me.yxp.qfun.utils.hook.HookUtils;
import me.yxp.qfun.utils.reflect.ClassUtils;
import me.yxp.qfun.utils.reflect.FieldUtils;
import me.yxp.qfun.utils.reflect.MethodUtils;

public final class OnAIOViewUpdate extends ApiHookItem {
    public static final ApiHookItem INSTANCE = new OnAIOViewUpdate();
    private static final String GRAY_TIPS_MSG_ITEM = "com.tencent.mobileqq.aio.msg.GrayTipsMsgItem";
    private static final String BUBBLE_LAYOUT_CLASS = "com.tencent.qqnt.aio.holder.template.BubbleLayoutCompatPress";
    private static final String HOOK_TAG = "HOOK";

    @Override
    public void loadHook() throws Throwable {
        Method sHandleUIState = MethodUtils.create(ClassUtils._AIOBubbleMsgItemVB())
                .withReturnType(void.class)
                .withParamTypes(ClassUtils.load("com.tencent.mvi.base.mvi.MviUIState"))
                .findOne();

        HookUtils.hookAlways(sHandleUIState, param -> {
            ViewGroup msgView = (ViewGroup) FieldUtils.create(param.thisObject)
                    .ofType(View.class)
                    .getValue();
            Object aIOMsgItem = FieldUtils.create(param.args[0]).ofType(ClassUtils._AIOMsgItem().getSuperclass()).getValue();

            if (aIOMsgItem == null || GRAY_TIPS_MSG_ITEM.equals(aIOMsgItem.getClass().getName())) {
                return;
            }

            Object msgRecord = FieldUtils.create(aIOMsgItem)
                    .ofType(ClassUtils._MsgRecord())
                    .inParent(ClassUtils._AIOMsgItem())
                    .getValue();

            LinearLayout msgLayout = findMsgLayout(msgView);
            if (msgLayout == null) {
                return;
            }

            FrameLayout frameLayout = setupFrameLayout(msgLayout);
            for (Listener listener : mListenerSet) {
                if (listener instanceof AIOViewUpdateListener) {
                    ((AIOViewUpdateListener) listener).onUpdate(frameLayout, msgRecord);
                }
            }
        }, null);
    }

    private LinearLayout findMsgLayout(ViewGroup msgView) {
        for (int i = 0; i < msgView.getChildCount(); i++) {
            View child = msgView.getChildAt(i);
            if (BUBBLE_LAYOUT_CLASS.equals(child.getClass().getName())) {
                return (LinearLayout) child;
            }
        }
        return null;
    }

    private FrameLayout setupFrameLayout(LinearLayout msgLayout) {
        FrameLayout frameLayout;

        if (msgLayout.getChildAt(0).getTag() == null
                || !HOOK_TAG.equals(msgLayout.getChildAt(0).getTag().toString())) {
            Context context = msgLayout.getContext();
            frameLayout = new FrameLayout(context);
            frameLayout.setTag(HOOK_TAG);

            View originalView = msgLayout.getChildAt(0);
            LinearLayout linearLayout = new LinearLayout(context);

            msgLayout.removeAllViews();
            linearLayout.addView(originalView);
            frameLayout.addView(linearLayout);
            msgLayout.addView(frameLayout);
        } else {
            frameLayout = (FrameLayout) msgLayout.getChildAt(0);
        }

        return frameLayout;
    }

    public interface AIOViewUpdateListener extends Listener {
        void onUpdate(FrameLayout frameLayout, Object msgRecord) throws Throwable;
    }
}

File: app/src/main/java/me/yxp/qfun/hook/api/OnAIOViewUpdate.kt
package me.yxp.qfun.hook.api

import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.LinearLayout
import androidx.core.view.children
import com.tencent.mobileqq.aio.msg.AIOMsgItem
import com.tencent.mobileqq.aio.msg.GrayTipsMsgItem
import com.tencent.mobileqq.aio.msglist.holder.AIOBubbleMsgItemVB
import com.tencent.mvi.base.mvi.MviUIState
import com.tencent.qqnt.aio.holder.template.BubbleLayoutCompatPress
import com.tencent.qqnt.kernel.nativeinterface.MsgRecord
import me.yxp.qfun.annotation.HookItemAnnotation
import me.yxp.qfun.hook.base.BaseApiHookItem
import me.yxp.qfun.hook.base.Listener
import me.yxp.qfun.utils.hook.hookAfter
import me.yxp.qfun.utils.reflect.findMethod
import me.yxp.qfun.utils.reflect.getObjectByType
import me.yxp.qfun.utils.reflect.getObjectByTypeOrNull

@HookItemAnnotation("监听消息视图更新")
object OnAIOViewUpdate : BaseApiHookItem<AIOViewUpdateListener>() {

    override fun loadHook() {

        AIOBubbleMsgItemVB::class.java
            .findMethod {
                returnType = void
                paramTypes(MviUIState::class.java)
            }.hookAfter(this) { param ->

                val parent = param.thisObject.getObjectByType<View>() as ViewGroup
                val aioMsgItem = param.args[0]
                    ?.getObjectByTypeOrNull(AIOMsgItem::class.java.superclass) as AIOMsgItem?
                if (aioMsgItem == null || aioMsgItem is GrayTipsMsgItem) return@hookAfter

                val msgRecord = aioMsgItem.msgRecord
                val msgLayout =
                    parent.children.filterIsInstance<BubbleLayoutCompatPress>().singleOrNull()
                        ?: return@hookAfter
                val frameLayout = setupFrameLayout(msgLayout)
                listenerSet.filter { verify(it) }
                    .forEach { it.onUpdate(frameLayout, msgRecord) }


            }
    }

    private fun setupFrameLayout(msgLayout: BubbleLayoutCompatPress): FrameLayout {

        val tag0 = msgLayout.getChildAt(0).tag

        if (tag0 == null || "HOOK" != tag0) {
            val context = msgLayout.context
            val origin = msgLayout.getChildAt(0)
            msgLayout.removeAllViews()
            val frameLayout = FrameLayout(context).apply {
                tag = "HOOK"
                addView(
                    LinearLayout(context).apply {
                        addView(origin)
                    }
                )
            }
            msgLayout.addView(frameLayout)
            return frameLayout

        } else {
            return msgLayout.getChildAt(0) as FrameLayout
        }
    }

}

fun interface AIOViewUpdateListener : Listener {
    fun onUpdate(frameLayout: FrameLayout, msgRecord: MsgRecord)
}